### ISSUE TITLE
broadcast randperm with dim specification

### ISSUE THREAD

**Issue by tchaton** [pytorch/pytorch#31679]
## 🚀 Feature

Currently randperm uses only int as input
It would be great to provide a tensor and specify a dimension to make a permutation on.

## Motivation

That would be an example of how to do right now
idx = torch.stack([torch.randperm(n) for _ in range(28*28)]).t().long() # permute pixel of an image
data_shuffled = torch.zeros(10, 784).scatter_(0, idx, data.view(10, -1)).view(10, 28, 28) # permute each image independently but in a vectorized way.

It isn't neither nice nor optimized.
